src/validation.cpp

```diff
@@ -4833,18 +4833,31 @@ void ChainstateManager::CheckBlockIndex()
         return;
     }
 
-    // Build forward-pointing map of the entire block tree.
+    // Build forward-pointing data structure for the entire block tree.
+    // For performance reasons, indexes of the best header chain are stored in a vector.
+    // All remaining blocks are stored in a multimap.
+    CBlockIndex* index_walk{m_best_header};
+    const int best_hdr_height{m_best_header->nHeight};
+    std::vector<CBlockIndex*> best_hdr_chain(best_hdr_height + 1);
+    while (index_walk) {
+        best_hdr_chain[index_walk->nHeight] = index_walk;
+        index_walk = index_walk->pprev;
+    }
+
     std::multimap<CBlockIndex*,CBlockIndex*> forward;
     for (auto& [_, block_index] : m_blockman.m_block_index) {
-        forward.emplace(block_index.pprev, &block_index);
+        // Only save indexes in forward that are not part of the best header chain.
+        if (block_index.nHeight > best_hdr_height || best_hdr_chain[block_index.nHeight]->GetBlockHash() != block_index.GetBlockHash()) {
```

Comment:
In commit "validation: improve performance of CheckBlockIndex" (0f7963e0b306257ae5d547ba979712156c9d3a30)

Calling GetBlockHash() here is unnecessary, right? It seems like it would be more consistent with rest of code to just compare pointers not hashes.

Reply:
Makes sense, but since I replaced then entire condition with `CChain::Contains()`, this is no longer relevant.